clipboard file: xrdp is not responding CB_RESPONSE_FAIL even if it failed

When `clipboard_send_file_data` fails due to the file is not readable or other reasons, the client waits to finish file transfer forever. Maybe we need to abort file transfer gracefully. 

![スクリーンショット 2021-09-15 15 36 12](https://user-images.githubusercontent.com/941609/133383004-0945ead2-7873-4ec0-b7e7-d6b2137bfa74.png)

```
meta@xrdp:~/xrdp$ tail -f ~/.local/share/xrdp/xrdp-chansrv.10.log
[20210915-06:35:14] [DEBUG] [process_message_channel_data(chansrv.c:446)] process_message_channel_data: chan_id 2 chan_flags 19
[20210915-06:35:14] [DEBUG] [clipboard_data_in(clipboard.c:1523)] clipboard_data_in: chan_id 2 chan_flags 0x13 length 32 total_length 32 in_request 0 g_ins->size 8192
[20210915-06:35:14] [DEBUG] [clipboard_data_in(clipboard.c:1586)] clipboard_data_in: clip_msg_id 8 clip_msg_status 0 clip_msg_len 24
[20210915-06:35:14] [DEBUG] [clipboard_data_in(clipboard.c:1591)] clipboard_data_in: 8
[20210915-06:35:14] [DEBUG] [clipboard_process_file_request(clipboard_file.c:508)] clipboard_process_file_request:
[20210915-06:35:14] [DEBUG] [clipboard_send_file_data(clipboard_file.c:408)] clipboard_send_file_data: streamId 24422 lindex 0 nPositionLow 0 cbRequested 34520
[20210915-06:35:14] [ERROR] [clipboard_send_file_data(clipboard_file.c:415)] clipboard_send_file_data: file open [/home/meta/Downloads/paper.docx] failed
[20210915-06:35:14] [DEBUG] [scard_pcsc_check_wait_objs(smartcard_pcsc.c:401)] scard_pcsc_check_wait_objs:
[20210915-06:35:14] [DEBUG] [scard_pcsc_get_wait_objs(smartcard_pcsc.c:374)] scard_pcsc_get_wait_objs:
[20210915-06:35:14] [DEBUG] [get_timeout(chansrv.c:153)] get_timeout:
```

I haven't read MS-RDPECLIP corner to corner yet however I guess xrdp should respond to [ File Contents Request PDU](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/cbc851d3-4e68-45f4-9292-26872a9209f2) with [File Contents Response PDU](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/df87c178-ab02-471a-acde-bb921aa1af85) that contain `CB_RESPONSE_FAIL` to indicate "the request could not be satisfied". 

I quickly performed `git grep CB_RESPONSE_FAIL` but it was not used in `clipboard_file.c` at all.  The return value of `clipboard_send_file_size` and `clipboard_send_file_data` should be checked and xrdp should return `CB_RESPONSE_FAIL` when they failed.

https://github.com/neutrinolabs/xrdp/blob/23742dbb8ef4f3a0b2d7d911bafd09153c6193da/sesman/chansrv/clipboard_file.c#L496-L526

--- a/sesman/chansrv/clipboard_file.c
+++ b/sesman/chansrv/clipboard_file.c
@@ -361,7 +361,8 @@ clipboard_send_file_size(int streamId, int lindex)
     }
     if (cfi->size < 0)
     {
-        LOG(LOG_LEVEL_ERROR, "clipboard_send_file_size: failed to get file size");
+        LOG(LOG_LEVEL_ERROR, "clipboard_send_file_size: error cfi->size is negative"
+            "value [%d]", cfi->size);
         clipboard_send_filecontents_response_fail(streamId);
         return 1;
     }
@@ -460,8 +461,8 @@ clipboard_send_file_data(int streamId, int lindex,
     }
     if (g_file_seek(fd, nPositionLow) < 0)
     {
-        LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_send_file_data: seek error "
-                  "in file: %s", full_fn);
+        LOG(LOG_LEVEL_ERROR, "clipboard_send_file_data: seek error in file [%s]: %s",
+            full_fn, g_get_strerror());
         g_file_close(fd);
         clipboard_send_filecontents_response_fail(streamId);
         return 1;

--- a/sesman/chansrv/clipboard_file.c
+++ b/sesman/chansrv/clipboard_file.c
@@ -82,6 +82,37 @@ timeval2wintime(struct timeval *tv)
 }
 #endif
 
+/***
+ * See MS-RDPECLIP 3.1.5.4.7
+ *
+ * Sends a failure response to a CLIPRDR_FILECONTENTS_REQUEST
+ * @param streamId Stream ID from CLIPRDR_FILECONTENTS_REQUEST
+ * @return 0 for success
+ */
+
+static int
+clipboard_send_filecontents_response_fail(int streamId)
+{
+    LOG_DEVEL(LOG_LEVEL_TRACE, "clipboardn_send_filecontents_response_fail:");
+
+    struct stream *s;
+    int size;
+    int rv;
+
+    make_stream(s);
+    init_stream(s, 64);
+
+    out_uint16_le(s, CB_FILECONTENTS_RESPONSE);
+    out_uint16_le(s, CB_RESPONSE_FAIL);
+    out_uint32_le(s, 4);
+    out_uint32_le(s, streamId);
+    s_mark_end(s);
+    size = (int)(s->end - s->data);
+    rv = send_channel_data(g_cliprdr_chan_id, s->data, size);
+    free_stream(s);
+    return rv;
+}
+
 /*****************************************************************************/
 /* this will replace %20 or any hex with the space or correct char
  * returns error */
@@ -318,12 +349,20 @@ clipboard_send_file_size(int streamId, int lindex)
     if (g_files_list == 0)
     {
         LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_send_file_size: error g_files_list is nil");
+        clipboard_send_filecontents_response_fail(streamId);
         return 1;
     }
     cfi = (struct cb_file_info *)list_get_item(g_files_list, lindex);
     if (cfi == 0)
     {
         LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_send_file_size: error cfi is nil");
+        clipboard_send_filecontents_response_fail(streamId);
+        return 1;
+    }
+    if (cfi->size < 0)
+    {
+        LOG(LOG_LEVEL_ERROR, "clipboard_send_file_size: failed to get file size");
+        clipboard_send_filecontents_response_fail(streamId);
         return 1;
     }
     file_size = cfi->size;
@@ -397,12 +436,14 @@ clipboard_send_file_data(int streamId, int lindex,
     if (g_files_list == 0)
     {
         LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_send_file_data: error g_files_list is nil");
+        clipboard_send_filecontents_response_fail(streamId);
         return 1;
     }
     cfi = (struct cb_file_info *)list_get_item(g_files_list, lindex);
     if (cfi == 0)
     {
         LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_send_file_data: error cfi is nil");
+        clipboard_send_filecontents_response_fail(streamId);
         return 1;
     }
     LOG_DEVEL(LOG_LEVEL_DEBUG, "clipboard_send_file_data: streamId %d lindex %d "
@@ -412,8 +453,9 @@ clipboard_send_file_data(int streamId, int lindex,
     fd = g_file_open_ex(full_fn, 1, 0, 0, 0);
     if (fd == -1)
     {
-        LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_send_file_data: file open [%s] failed",
-                  full_fn);
+        LOG(LOG_LEVEL_ERROR, "clipboard_send_file_data: file open [%s] failed: %s",
+            full_fn, g_get_strerror());
+        clipboard_send_filecontents_response_fail(streamId);
         return 1;
     }
     if (g_file_seek(fd, nPositionLow) < 0)
@@ -421,6 +463,7 @@ clipboard_send_file_data(int streamId, int lindex,
         LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_send_file_data: seek error "
                   "in file: %s", full_fn);
         g_file_close(fd);
+        clipboard_send_filecontents_response_fail(streamId);
         return 1;
     }
     make_stream(s);
@@ -432,6 +475,7 @@ clipboard_send_file_data(int streamId, int lindex,
                   cbRequested, size);
         free_stream(s);
         g_file_close(fd);
+        clipboard_send_filecontents_response_fail(streamId);
         return 1;
     }
     out_uint16_le(s, CB_FILECONTENTS_RESPONSE); /* 9 */

--- a/sesman/chansrv/clipboard_file.h
+++ b/sesman/chansrv/clipboard_file.h
@@ -35,6 +35,7 @@ clipboard_c2s_in_files(struct stream *s, char *file_list);
 
 int
 clipboard_request_file_size(int stream_id, int lindex);
+
 int
 clipboard_request_file_data(int stream_id, int lindex, int offset,
                             int request_bytes);

